azurerm/internal/services/compute/linux_virtual_machine_resource.go

```diff
@@ -993,6 +1001,57 @@ func resourceLinuxVirtualMachineUpdate(d *schema.ResourceData, meta interface{})
 			return fmt.Errorf("Once a customer-managed key is used, you can’t change the selection back to a platform-managed key")
 		}
 	}
+	deleteRemovedDisks := meta.(*clients.Client).Features.VirtualMachine.DeleteDataDiskOnDeletion
+	dataDisksToDeleted := make([]compute.DataDisk, 0)
+	// TODO Beta flag here
+	if true {
+		if d.HasChange("data_disk") {
+			shouldUpdate = true
+			updatedDataDisks := expandVirtualMachineDataDisks(d.Get("data_disk").([]interface{}))
+			dataDisks := make([]compute.DataDisk, 0)
+			// Reconcile the data disks if previously specified
+			if existing.VirtualMachineProperties.StorageProfile != nil && existing.VirtualMachineProperties.StorageProfile.DataDisks != nil {
+				// Because some values are computed we need to use the existing disk data or a change of order will break values such as `managed_disk_id`
+				// This also filters out removed disks
+				for _, existingDataDisk := range *existing.VirtualMachineProperties.StorageProfile.DataDisks {
+					found := false
+					for _, dataDisk := range *updatedDataDisks {
+						if *dataDisk.Name == *existingDataDisk.Name {
```

Comment:
can we ignore nil checks here?

Reply:
Partially, yes. The other was an oversight, thanks :)